polynote-kernel/src/main/scala/polynote/messages/Messages.scala

```diff
@@ -127,47 +125,175 @@ object RunCell extends MessageCompanion[RunCell](3)
 final case class CellResult(notebook: ShortString, id: TinyString, result: Result) extends Message
 object CellResult extends MessageCompanion[CellResult](4)
 
-/**
-  * A ContentEdit at a position deletes some amount (possibly nothing) after that position, and then inserts some
-  * content (possibly empty) at that position.
-  *
-  * @param pos          The position at which the edit occurs
-  * @param deleteLength The number of characters to delete after the given position
-  * @param content      The content to insert
-  */
-final case class ContentEdit(pos: Int, deleteLength: Int, content: String) {
-  def applyTo(rope: Rope): Rope = if (deleteLength > 0) rope.delete(pos, deleteLength).insertAt(pos, Rope(content)) else rope.insertAt(pos, Rope(content))
 
-  // Given another edit which occured after this edit, create an edit that will be equivalent to applying this edit
-  // when applied after the given edit instead.
-  def rebase(other: ContentEdit): ContentEdit = other match {
-    // if the other edit is entirely before this edit, just shift this edit by the length delta
-    case ContentEdit(otherPos, otherDeleteLength, otherContent) if otherPos + otherDeleteLength <= pos =>
-      copy(pos = pos + otherContent.length - otherDeleteLength)
+sealed trait ContentEdit {
+  def pos: Int
+  def lengthDelta: Int
 
-    // if the other edit is entirely after this edit, nothing to do
-    case ContentEdit(otherPos, _, _) if otherPos >= pos + deleteLength => this
+  def rebase(other: ContentEdits): ContentEdits = ContentEdits(ShortList(ContentEdit.rebaseAll(this, other.edits)._1))
 
-    // if the other edit affects the region that was replaced by this edit, then they conflict
-    // we want the minimal edit that replaces the entire conflict region
-    case ContentEdit(otherPos, otherDeleteLength, otherContent) =>
+  def applyTo(rope: Rope): Rope = this match {
+    case Insert(pos, content) => rope.insertAt(pos, Rope(content))
+    case Delete(pos, length)  => rope.delete(pos, length)
+  }
 
-      // start at the earlier position
-      val start = math.min(pos, otherPos)
+  def nonEmpty: Boolean
+}
 
-      // They deleted to this position
-      val theirTarget = otherPos + otherDeleteLength
+final case class Insert(pos: Int, content: String) extends ContentEdit {
+  def lengthDelta: Int = content.length
 
-      // I would have deleted up to this position...
-      val myTarget = pos + deleteLength
+  def nonEmpty: Boolean = content.nonEmpty
 
-      // So I should delete this much extra
-      val deleteExtra = math.max(0, myTarget - theirTarget) + math.max(0, otherPos - pos)
+  override def toString: String = s"""Insert($pos, "$content")"""
+}
 
-      val combinedContent = if (otherPos <= pos) otherContent + content else content + otherContent
+object Insert {
+  implicit val insertTag: Discriminator[ContentEdit, Insert, Byte] = Discriminator(0)
+}
+
+final case class Delete(pos: Int, length: Int) extends ContentEdit {
```

Comment:
I think both pos and length can only be positive, right (same goes for Insert pos I think)? In that case maybe we could add some validation to ensure that we don't accidentally create any edits with negative values, just as a sanity check. 

Since we do lots of arithmetic with these values I think things could break pretty badly if any negative numbers get somehow added to the mix, in which case I bet it would break in weird ways that would be difficult to chase down, so we might want to fail fast?